force-app/main/default/classes/ucl_NightlyService.cls

```diff
@@ -0,0 +1,256 @@
+
+  private static void upsertLegalAdvisorRecords(
+    List<LegalAdvisorResponse> legalAdvisors
+  ) {
+    Set<Legal_Advisor__c> legalAdvisorRecords = new Set<Legal_Advisor__c>();
+
+    Set<String> legalAdvisorAccountNumbers = new Set<String>();
+    for (LegalAdvisorResponse legalAdvisor : legalAdvisors) {
+      legalAdvisorAccountNumbers.add(legalAdvisor.AccountNumber);
+    }
+
+    // Query existing legal advisors based on the account numbers in the input list
+    Set<String> existingLegalAdvisors = new Set<String>();
+    for (Legal_Advisor__c existingLegalAdvisor : [
+      SELECT Account_Number__c
+      FROM Legal_Advisor__c
+      WHERE Account_Number__c IN :legalAdvisorAccountNumbers
+    ]) {
+      existingLegalAdvisors.add(existingLegalAdvisor.Account_Number__c);
+    }
+
+    for (LegalAdvisorResponse legalAdvisor : legalAdvisors) {
+      Boolean legalAdvisorExists = existingLegalAdvisors.contains(
+        legalAdvisor.AccountNumber
+      );
+      if (legalAdvisor.AccountStatus == 'Disabled') {
+        if (legalAdvisorExists) {
+          // if Account Status is now Disabled but the record already exists in the org
+          Legal_Advisor__c newLegalAdvisor = createLegalAdvisorRecord(
+            legalAdvisor
+          );
+          legalAdvisorRecords.add(newLegalAdvisor);
+        }
+      } else if (legalAdvisor.AccountStatus == 'Enabled') {
+        // if Account Status is Enabled
+        Legal_Advisor__c newLegalAdvisor = createLegalAdvisorRecord(
+          legalAdvisor
+        );
+        legalAdvisorRecords.add(newLegalAdvisor);
+      }
+    }
+    List<Legal_Advisor__c> legalAdvisorList = new List<Legal_Advisor__c>(legalAdvisorRecords);
+
+    upsert legalAdvisorList Account_Number__c;
+  }
+
+  private static Legal_Advisor__c createLegalAdvisorRecord(
+    LegalAdvisorResponse legalAdvisor
+  ) {
+    Legal_Advisor__c newLegalAdvisor = new Legal_Advisor__c();
+
+    newLegalAdvisor.Account_Name__c = legalAdvisor.AccountName;
+    newLegalAdvisor.Name = legalAdvisor.AccountName;
+    newLegalAdvisor.Account_Number__c = legalAdvisor.AccountNumber;
+    newLegalAdvisor.Account_Status__c = legalAdvisor.AccountStatus;
+    newLegalAdvisor.As_of_Date__c = Date.parse(legalAdvisor.AsOfDate);
+    return newLegalAdvisor;
+  }
+
+  // CLIENTS
+  private static Quiddity uploadedBy;
+  private static void upsertClientRecords(
+    List<LegalAdvisorResponse> legalAdvisors
+  ) {
+    uploadedBy = Request.getCurrent().getQuiddity(); // know if the client is being uploaded by an admin in the Clients record page
+    List<Client__c> newClientRecords = new List<Client__c>();
+
+    //fetch client numbers that already exist in the org
+    List<Client__c> existingClients = new List<Client__c>();
+    for (Client__c existingClient : [
+      SELECT
+        Client_Number__c,
+        First_Name__c,
+        Last_Name__c,
+        Email__c,
+        Phone__c,
+        Address__c,
+        External_Id__c
+      FROM Client__c
```

Comment:
be careful with queries without filtering data, please use WHERE clause or LIMIT BY